genawaiter-proc-macro/src/common.rs

```diff
@@ -0,0 +1,211 @@
+use std::collections::VecDeque;
+
+use proc_macro2::TokenStream as TokenStream2;
+use proc_macro_error::abort;
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream},
+    parse2,
+    spanned::Spanned,
+    visit_mut::{self, VisitMut},
+    Block,
+    Expr,
+    ExprClosure,
+    Item,
+    Macro,
+    Result as SynResult,
+    Stmt,
+    Token,
+    Type,
+};
+
+pub struct YieldMatchMacro {
+    pub parent: Option<Stmt>,
+    pub collected: Vec<Option<Stmt>>,
+    pub coll_replace: Vec<Stmt>,
+}
+
+impl YieldMatchMacro {
+    pub fn new() -> Self {
+        Self {
+            parent: None,
+            collected: vec![],
+            coll_replace: vec![],
+        }
+    }
+}
+
+impl VisitMut for YieldMatchMacro {
+    fn visit_stmt_mut(&mut self, node: &mut Stmt) {
```

Comment:
This is a big improvement over the last implementation! Unfortunately there are still more cases it doesn't cover, such as calling a method on the resume argument:

<details>
<summary>Details</summary>

```rust
#[test]
fn sync_proc_macro_fn() {
    #[genawaiter::sync::sync_yield_fn(u8)]
    async fn odds() {
        for n in (1_u8..).step_by(2).take_while(|&n| n < 10) {
            let cloned_resume_arg = genawaiter::yield_!(n).clone();
        }
    }
    let gen = Gen::new(odds);
    let res = gen.into_iter().collect::<Vec<_>>();
    assert_eq!(vec![1, 3, 5, 7, 9], res)
}
```

</details>

I think to cover all your bases you'll need to think in terms of expressions (`Expr`) instead of `Stmt`s.

Reply:
I got to remove even more code sweet. That worked really well.